File: source/downloads/code/LeetCode/Java/Reverse-Linked-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *      int val;
 *      ListNode next;
 *      ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        return reverse(head,null);           
    }

    public ListNode reverse(ListNode head, ListNode newhead){
        if(head==null)
            return newhead;
        ListNode nextNode = head.next;
        head.next = newhead;
        return reverse(nextNode,head);
        
    }
}

File: downloads/code/Reverse-Linked-List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        if head == None or head.next == None:
            return head

        prev = ListNode(-1)
        prev.next = head

        node = head

        while node:
            nextnode = node.next
            if node == head:
                node.next = None
            else:
                node.next = prev
            prev = node
            node = nextnode
        return prev
            
